holopy/core/io/io.py

```diff
@@ -413,31 +414,69 @@ def load_average(filepath, refimg=None, spacing=None, medium_index=None, illum_w
     if channel is None and refimg is not None and illumination in refimg.dims:
         channel = [i for i, col in enumerate(['red','green','blue']) if col in refimg[illumination].values]
 
-    accumulator = clean_concat([load_image(image, spacing, channel=channel) for image in filepath],'images')
-
     if np.isscalar(spacing):
         spacing = np.repeat(spacing, 2)
 
+    # calculate the average
+    accumulator = Accumulator()
+    for path in filepath:
+        accumulator.push(load_image(path, spacing, channel=channel))
+    mean_image = accumulator.mean()
+
+    # calculate average noise from image
+    if noise_sd is None and len(filepath) > 1:
+        noise_sd = ensure_array(accumulator.cv())
+
     # crop according to refimg dimensions
     if refimg is not None:
         def extent(i):
             name = ['x','y'][i]
             return np.around(refimg[name].values/spacing[i]).astype('int')
-        accumulator = accumulator.isel(x=extent(0), y=extent(1))
-        accumulator['x'] = refimg.x
-        accumulator['y'] = refimg.y
-
-    # calculate the average
-    mean = accumulator.mean('images')
-
-    # calculate average noise from image
-    if noise_sd is None and len(filepath) > 1:
-        noise_sd = ensure_array((accumulator.std('images')/mean).mean(('x','y','z')))
-    accumulator = mean
+        mean_image = mean_image.isel(x=extent(0), y=extent(1))
+        mean_image['x'] = refimg.x
+        mean_image['y'] = refimg.y
 
     # copy metadata from refimg
     if refimg is not None:
-        accumulator = copy_metadata(refimg, accumulator, do_coords=False)
+        mean_image = copy_metadata(refimg, mean_image, do_coords=False)
 
     # overwrite metadata from refimg with provided values
-    return update_metadata(accumulator, medium_index, illum_wavelen, illum_polarization, normals, noise_sd)
+    return update_metadata(mean_image, medium_index, illum_wavelen, illum_polarization, normals, noise_sd)
+
+
+class Accumulator:
+    """Calculates average and coefficient of variance for numerical data in
+    one pass using Welford's algorithim.
+    """
+    def __init__(self):
+        self._n = 0
+        self._running_mean = None
+        self._running_var = None
+
+    def push(self, x):
+        self._n += 1
+
+        if self._n == 1:
+            self._running_var = x * 0.0
+            self._running_mean = self._running_var + x
+        else:
+            self._running_var += ((x - self._running_mean) *
+             ((x - (self._running_mean + (x - self._running_mean) / self._n))))
+            self._running_mean += (x - self._running_mean) / self._n
+
+    def mean(self):
+        return self._running_mean if self._running_mean is not None else 0.0
+
+    def cv(self):
+        """ The coefficient of variation
+        """
+        if self._n == 0:
+            return None
+        else:
+            return np.mean(np.array(self._std() / self.mean()))
```

Comment:
This won't work for multicolour images. Previously, the mean was taken only over dimensions {x,y,z} but now it's been replaced with np.mean which averages over all dimensions.

Reply:
Is the color axis dimesion 0 or -1?

Reply:
Good catch @barkls . @barkls since you are the one who primarily uses the 2-color hologram, can I ask you to write a failing test? I'm worried that if @ralex0 writes it, the test will be incomplete, and there is no point in having a test that doesn't actually catch the desired use case.

Reply:
New test is in and passing now.